pyroomacoustics/room.py

```diff
@@ -2165,137 +2202,450 @@ def compute_rir(self):
+                    print("Minimum Time", time.min() * self.fs)
 
                 else:
                     t_max = 0.0
 
                 if self.simulator_state["rt_needed"]:
 
                     # get the maximum length from the histograms
-                    nz_bins_loc = np.nonzero(self.rt_histograms[m][s][0].sum(axis=0))[0]
+                    nz_bins_loc = np.nonzero(self.rt_histograms[m][s][0].sum(axis=0))[
+                        0
+                    ]  # Sum vertically across octave band for each value in histogram (7,2500) -> (2500) -> np .nonzero(
+
                     if len(nz_bins_loc) == 0:
                         n_bins = 0
                     else:
                         n_bins = nz_bins_loc[-1] + 1
 
                     t_max = np.maximum(t_max, n_bins * self.rt_args["hist_bin_size"])
 
+                    # N changes here , the length of RIR changes if we apply RT method.
                     # the number of samples needed
                     # round up to multiple of the histogram bin size
                     # add the lengths of the fractional delay filter
+
                     hbss = int(self.rt_args["hist_bin_size_samples"])
+
                     N = int(math.ceil(t_max * self.fs / hbss) * hbss)
 
-                # this is where we will compose the RIR
+                # this is where we will compose the RIR i.e length of RIR
                 ir = np.zeros(N + fdl)
 
                 # This is the distance travelled wrt time
                 distance_rir = np.arange(N) / self.fs * self.c
 
                 # this is the random sequence for the tail generation
                 seq = sequence_generation(volume_room, N / self.fs, self.c, self.fs)
-                seq = seq[:N]
+                seq = seq[:N]  # take values according to N as seq is larger
 
                 # Do band-wise RIR construction
                 is_multi_band = self.is_multi_band
                 bws = self.octave_bands.get_bw() if is_multi_band else [self.fs / 2]
                 rir_bands = []
 
-                for b, bw in enumerate(bws):
+                """
+                Use octave bands to construct RIR :
+                1) Ray-tracing is activated
+                2) directivity of both the microphones and source is not given.
 
-                    ir_loc = np.zeros_like(ir)
+                """
+                # Improve the checks.
 
-                    # IS method
-                    if self.simulator_state["ism_needed"]:
+                if (
+                    (
+                        self.mic_array.directivity is None
+                        or isinstance(self.mic_array.directivity[m], CardioidFamily)
+                    )
+                    and (
+                        self.sources[s].directivity is None
+                        or isinstance(self.sources[s].directivity, CardioidFamily)
+                    )
+                ) or self.simulator_state["rt_needed"]:
```

Comment:
why is the last condition needed ?

Reply:
It may be easier to order this if clause the other way around. Now, there are only two ways to proceed:
1) If either mic or source is from DIRPAT class, we use the new way
2) If not we use the other way
Thus
```python
if isinstance(self.mic_array.directivity[m], DIRPATPattern) or isinstance(self.sources[s].directivity, DIRPATPattern):
    # new way
else:
    # old way
```